pkg/dockerregistry/server/util.go

```diff
@@ -0,0 +1,84 @@
+package server
+
+import (
+	"fmt"
+
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/context"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	kerrors "k8s.io/kubernetes/pkg/api/errors"
+
+	imageapi "github.com/openshift/origin/pkg/image/api"
+)
+
+const (
+	// EventReasonPushDenied is a name of reason for denial of blob write or manifest put. May be used by
+	// event listeners to filter the events on image stream objects.
+	EventReasonPushDenied = "PushDenied"
+)
+
+// autoProvisionImageStream attempts to create an image stream for given repository in etcd. An existence
+// check shell done before calling this function.
+func autoProvisionImageStream(ctx context.Context, repo *repository) (*imageapi.ImageStream, error) {
+	stream := &imageapi.ImageStream{
+		ObjectMeta: kapi.ObjectMeta{
+			Name: repo.name,
+		},
+	}
+
+	client, ok := UserClientFrom(ctx)
+	if !ok {
+		return nil, fmt.Errorf("Error creating user client to auto provision image stream: Origin user client unavailable")
+	}
+
+	stream, err := client.ImageStreams(repo.namespace).Create(stream)
+	if err != nil {
+		statusErr, ok := err.(*kerrors.StatusError)
+		if !ok {
+			return nil, fmt.Errorf("Error auto provisioning image stream: %v", err)
+		}
+		if kerrors.IsAlreadyExists(statusErr) {
+			context.GetLogger(ctx).Infof("Cannot create image stream %s/%s that already exists", repo.namespace, repo.name)
+			return client.ImageStreams(repo.namespace).Get(repo.name)
+		}
+		if kerrors.IsForbidden(statusErr) {
+			context.GetLogger(ctx).Errorf("Denied auto provisioning image stream: %v", statusErr)
+			reportPushDeniedEvent(ctx, repo, nil, statusErr.Error())
```

Comment:
Since `reportPushDeniedEvent` will attempt to create the image stream when the 3rd function argument is `nil`, and `reportPushDeniedEvent` calls this function, will this put you in an endless loop?